utils/R/convert.input.R

```diff
@@ -6,65 +6,144 @@
 ##' @export
 ##' @author Betsy Cowdery, Michael Dietze, Ankur Desai
 convert.input <- function(input.id, outfolder, formatname, mimetype, site.id, start_date, end_date, 
-                          pkg, fcn, username, con=con, host, browndog, write=TRUE,  
-                          format.vars=format.vars, ...) {
-  l <- list(...); #print(l)
+                          pkg, fcn, con=con, host, browndog, write=TRUE,  
+                          format.vars, overwrite=FALSE, ...) {
+  input.args <- list(...)
+
+  logger.debug(paste(
+    "Convert.Inputs", fcn, input.id, host$name, outfolder, formatname, mimetype, 
+    site.id, start_date, end_date))
 
-  logger.info(paste("Convert.Inputs",fcn,input.id,host$name,outfolder,formatname,mimetype,site.id,start_date,end_date))
   n <- nchar(outfolder)
-  if(substr(outfolder,n,n) != "/"){outfolder = paste0(outfolder,"/")}
-  
-  outname = tail(unlist(strsplit(outfolder,'/')),n=1)
+  if(substr(outfolder,n,n) != "/") {
+    outfolder = paste0(outfolder,"/")
+  }
   
-  print(start_date)
-  startdate <- as.POSIXlt(start_date, tz = "GMT")
-  enddate   <- as.POSIXlt(end_date, tz = "GMT")
+  outname = tail(unlist(strsplit(outfolder,'/')), n=1)
 
-  # Consider adding a force option to skip the check and continue with conversion
-  print("start CHECK")
-  check = dbfile.input.check(site.id, startdate, enddate, mimetype, formatname, parentid=input.id, con=con, host$name)
+  print(paste(
+    "start CHECK Convert.Inputs", fcn, input.id, host$name, outfolder, formatname, mimetype, 
+    site.id, start_date, end_date))
+  existing.dbfile <- dbfile.input.check(
+    siteid=site.id, mimetype=mimetype, formatname=formatname, parentid=input.id, 
+    con=con, hostname=host$name, ignore.dates=TRUE)
+  print(existing.dbfile, digits=10)
   print("end CHECK")
-  print(check)
-  if(length(check)>0){
-    return(list(input.id=check$container_id, dbfile.id=check$id))
+
+  if(nrow(existing.dbfile) > 0) {
+    if(nrow(existing.dbfile) > 1) {
+      print(existing.dbfile)
+      logger.warn("Multiple existing inputs found. Using last.")
+      existing.dbfile <- existing.dbfile[nrow(existing.dbfile),]
+    }
+
+    existing.input <- db.query(paste0("SELECT * FROM inputs WHERE id=", existing.dbfile[['container_id']]), con)    
+    # Convert dates to Date objects and strip all time zones (DB values are timezone-free)
+    start_date <- force_tz(as_date(start_date), 'GMT')
+    end_date <- force_tz(as_date(end_date), 'GMT')
+    existing.input$start_date <- force_tz(as_date(existing.input$start_date), 'GMT')
+    existing.input$end_date <- force_tz(as_date(existing.input$end_date), 'GMT')
+
+    if(overwrite) {
+      # start and end dates stay the same. 
+      # collect files to flag for deletion
+      existing.dbfile <- dbfile.check('Input', existing.input[['id']], con, host$name)
+      if(nrow(existing.dbfile) > 0) {
```

Comment:
By this point convert.input knows there is an existing input record and dbfile record, and overwrite is TRUE. So there are going to be files to delete. But it needs to wait until it know it's been successful. So, here it collects the file paths to be deleted, and creates string versions of the remote functions needed to (1) move them to a temp location, (2) delete the temp location (if successful) and/or (3) replace them to their original locations (if unsuccessful.

Some limitations:
1. If convert.input is partially successful, you could end up with some new (but presumably unwanted) files, mixed in with the old ones that were replaced on exiting this failing function.
2. I don't actually check to make sure that the remote function successfully executes. I think this is probably ok in most places, because if it doesn't return a `result` object with a valid `result$file` then this function should error out, triggering the replace-old-files routine. But that needs some testing.
3. I meant to but haven't added a mechanism to delete unused dbfile entries. Normally I think the overwriting call to convert.inputs will be reusing the same input and dbfile records, so it's no problem. But for the case where you have a dbfile record pointing to one location, and now decide to write to another, we may want to make sure to obliterate the obsolete one?